sound/soc/sof/sof-client-probes-ipc4.c

```diff
@@ -0,0 +1,276 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2019-2022 Intel Corporation. All rights reserved.
+//
+// Author: Jyri Sarha <jyri.sarha@intel.com>
+//
+
+#include <sound/soc.h>
+#include <sound/sof/ipc4/header.h>
+#include <uapi/sound/sof/header.h>
+#include "sof-priv.h"
+#include "ipc4-priv.h"
+#include "sof-client.h"
+#include "sof-client-probes.h"
+#include "ipc4-topology.h"
+
+enum sof_ipc4_dma_type {
+	SOF_IPC4_DMA_HDA_HOST_OUTPUT = 0,
+	SOF_IPC4_DMA_HDA_HOST_INPUT = 1,
+	SOF_IPC4_DMA_HDA_LINK_OUTPUT = 8,
+	SOF_IPC4_DMA_HDA_LINK_INPUT = 9,
+	SOF_IPC4_DMA_DMIC_LINK_INPUT = 11,
+	SOF_IPC4_DMA_I2S_LINK_OUTPUT = 12,
+	SOF_IPC4_DMA_I2S_LINK_INPUT = 13,
+};
+
+union sof_ipc4_connector_node_id {
+	u32 val;
+	struct {
+		u32 vindex:8;
+		u32 dma_type:5;
+		u32 rsvd:19;
```

Comment:
for the sake of consistency in the SOF driver, I insist @jsarha :)

Reply:
Ok, I will do it then. However, I still think it does not help with the code readability, and the connector_node_id starts to look dangerously similar to IPC3 buffer_id.

Reply:
@jsarha, what do you mean (and found) regarding to `DMA I should use the already existing macros for IPC4 node_id`?
_That_ `node_id` in `struct sof_copier_gateway_cfg` is a completely different (for convenience) than the `node_id` for the probes's `struct sof_ipc4_probe_gtw_cfg`

I think what would be nicer to do is to get rid of the union, use u32 and macros to shift things into place as we do in other places in SOF kernel code.

Reply:
Ok, its not at all trivial to draw the connections between payload struct data members and macros that should be used to fill the bit-fields within those members. Not at all convinced this is better system. I was fooled that a node_id is a node_id,  and same macros can be used anywhere since the naming and bit-fields match.  Well,  I try to do better job with IPC4 PROBES macros.